force-app/main/default/classes/DMLOperations.cls

```diff
@@ -179,28 +234,58 @@ public with sharing class DMLOperations {
 	 * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
 	 */
 	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
+        list<String> contactLastname = new List<String>();
 		// Loop through each Contact
-
-		// Extract the Account Name from Contact's LastName
-
-		// Get the account based on the name or create one if it doesn't exist
-
+        for(Contact currContact : contactsToUpsert) {
+            contactLastname.add(currContact.LastName);
+        }
+        // Query for Account Id based on Contact's lastname
+        // Extract the Account Name from Contact's LastName
+        map<String, Account> existingAccountsMap = new map<string, Account>();
+        List<Account> existingAccountList = [SELECT Id, Name FROM Account WHERE Name IN :contactLastname];
+        for (Account currAcc : existingAccountList) {
+            existingAccountsMap.put(currAcc.Name, currAcc);
+        }
+       // Get the account based on the name or create one if it doesn't exist
+        for (String currContact : contactLastname) {
+            if (existingAccountsMap.get(currContact) == null) {
+                Account newAccount = upsertAccount(currContact);
+                existingAccountsMap.put(currContact, newAccount);
+            }
+        }
 		// Link the Contact to its Account using the Account Id
-
+        for(Contact currContact : contactsToUpsert) {
+            currContact.AccountId = existingAccountsMap.get(currContact.LastName).Id;
+        }
 		// Upsert the Contacts
-	}
-
+         upsert contactsToUpsert;
```

Comment:
Great work on this one! Its tricky to fully avoid any DML or queries in the loops